semantics/psc-trees.ads

```diff
@@ -99,6 +109,9 @@ package PSC.Trees is
       New_Operand : Optional_Tree);
    --  Set Nth operand of given Tree
 
+   function Kind (T : Tree) return Tree_Kind_Enum;
```

Comment:
You might declare this as an abstract function.  This will ensure that it is always overridden in all descendants.  It also means there is no need to create a "dummy" body with an Assert(False).